Fix dock hover wobble #2

nmfretz opened the pull request:
This pull request proposes changes that fix the dock wobble issue noted by @Sun0fABeach in Issue #434, while preserving the margins within the dock. This works on both desktop and mobile (when the dock is mounted to the left). 

https://user-images.githubusercontent.com/85373263/193135690-f97b9568-f880-4f2b-84d8-c157572b0a66.mp4

I tried not to change the underlying html or javascript too much, and instead focused on a css solution. This could work as a short-term solution to the dock wobble issue; however, I think a longer-term solution would be to rewrite some of the html in `Dock.vue` and `DockApp.vue` so that the css can be simplified.

**Some notes about the changes (not exhaustive):**

- Replaced margin property with padding for .dock-app-container. Removed height/width properties from .dock-app-container and added them to .dock-app-icon (and dark/light-mode svg). These two changes result in the user always being in a hover state when moving from one dock-app to the next because the cursor never enters margin space between the apps.

- .dock-app-icon (and dark/light-mode svg) change size on hover instead of .dock-app-icon-container. This is still triggered when the user hovers over .dock-app-container. Padding property keeps appropriate spacing within the dock.

- Changed the dock divider (that separates the settings app and dark/light-mode toggle) from a `<div>` to a pseudo ::after element on the settings app. This prevents the user from exiting out of a hover state when moving between the settings app and the dark/light-mode toggle.

  - bounce-vertical-reverse and bounce-horizontal-reverse animations are applied to the ::after element so that it does not move when the settings app is clicked and bounces.

These changes unfortunately mean that the .dock-app-notification `<div>` needs to be scaled separately from the .dock-app-icon-container when a user hovers over .dock-app-container.  With my changes, I am not entirely happy with how the transition looks when a user hovers over the app-store icon that has a notification for app updates. Needs some work.

What do you think? It may be a bit too hacky for a merge, but if enough people complain about the dock it could be an interim fix.

Sun0fABeach commented:
Great work, feels much better now!

nmfretz commented:
https://user-images.githubusercontent.com/85373263/193134673-39444eb9-fdae-4348-bf07-bd97fc737927.mp4

nmfretz commented:
> Great work, feels much better now!

Thanks Markus! There's still a little bit of tweaking that could be done I think. At this point I may have just added too many css work-arounds.

mayankchhabra commented on src/components/DockApp.vue:
```diff
@@ -188,26 +248,58 @@ export default {
-      .dock-app-name  {
+      .dock-app-icon-container {
+        .dock-app-notification {
+          transform: scale(1.3) translate(-1px, -17px);
```
Shall we use `scale3d` and `translate3d` here for GPU acceleration?

nmfretz replied:
Yes. I have now updated all instances where I used `scale` and `transform` to `scale3d` and `translate3d`. I should have clued in to this from your last commit to `DockApp.vue`.

mayankchhabra commented on src/components/DockApp.vue:
```diff
@@ -188,26 +248,58 @@ export default {
-        transform: scale3d(1.3, 1.3, 1.3) translate3d(10px, 0, 0);
+        .dock-app-icon-container {
+          .dock-app-notification {
+            transform: scale(1.3) translate(20px, 1px);
```
And use `scale3d` and `translate3d` here as well for GPU acceleration?

nmfretz replied:
Done

mayankchhabra reviewed (commented):
Thanks for taking this up, @nmfretz! Impressive work and really clever workarounds, especially on using padding and changing divider to a pseudo element. Hovering over the icons feels very elegant.

>  These two changes result in the user always being in a hover state when moving from one dock-app to the next because the cursor never enters margin space between the apps.

Very clever and it also feels much more intuitive this way because the moment your cursor hovers near a dock icon, it inflates in very dock-esque manner.

One small issue I noticed: the position of active state indicator and notification on the settings icon seems off:

<img width="412" alt="image" src="https://user-images.githubusercontent.com/10330103/193332442-d55af6d1-c003-42bc-ac90-494340d11dfa.png">

Also, can you elaborate a bit more on this:
> With my changes, I am not entirely happy with how the transition looks when a user hovers over the app-store icon that has a notification for app updates. Needs some work.

Your PR:

https://user-images.githubusercontent.com/10330103/193332685-59c2ea6f-4549-4bdb-b699-7bc64e3ceca3.mp4

Current:

https://user-images.githubusercontent.com/10330103/193332891-2adcc82d-afb7-4d7c-85f3-6b746e291686.mp4

Are you referring to the parallax-like effect in the manner that the notification badge expands and contracts compared to the icon?

nmfretz commented:
Thanks for the kind words, Mayank. I have pushed a new commit that addresses the issues you noted.

> One small issue I noticed: the position of active state indicator and notification on the settings icon seems off:
> 
Good catch, that was a silly oversight on my part. The offset occurs because the extra padding on the psuedo :after element makes the settings app container wider. I had taken this into account for the app name tooltip, but forgot about the state and notification indicators. 

To avoid this problem entirely, I have changed the dock divider to be a pseudo :before element on the dark-light mode toggle, which doesn't have state or notification indicators.


> Are you referring to the parallax-like effect in the manner that the notification badge expands and contracts compared to the icon?
>
Yes, exactly. However, the mp4 you included of my PR running on your computer looks great. My current laptop is not very powerful and I think it was struggling a bit to serve the application and render the transformations, so I was sometimes seeing a very subtle offset in the expansion/contraction of the icon and notification badge.

mayankchhabra reviewed (approved):
Tested locally and everything looks great! Nice work here, @nmfretz. Thank you!

mayankchhabra commented:
> My current laptop is not very powerful and I think it was struggling a bit to serve the application and render the transformations

Just curious, what are the CPU/GPU/RAM specs of your laptop? A few users have reported laggy animations throughout the UI and I'm wondering if the spec-range is common between all reports and if there's a way we can make the experience better...

nmfretz commented:
> Just curious, what are the CPU/GPU/RAM specs of your laptop? A few users have reported laggy animations throughout the UI and I'm wondering if the spec-range is common between all reports and if there's a way we can make the experience better...

Current specs of my Surface Laptop 2:
- **Processor**: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz 2.11 GHz
- **RAM**: 8 GB
- **GPU**: Intel(R) UHD Graphics 620

My animations are a bit laggy.
